pkg/utils/utils.go

```diff
@@ -310,3 +310,27 @@ func IsValidCustomGPUArray(s string) bool {
 
 	return true
 }
+
+// CheckEnvsToRemove checks the current envs against the desired ones and returns the envs that needs to be
+// removed from the target env list and envs that need to be kept.
+// Returns envsToRemove, envsToKeep
+func CheckEnvsToRemove(desired, current []v1.EnvVar) ([]v1.EnvVar, []v1.EnvVar) {
```

Comment:
i am thinking.
so why not just build the `"$patch": "delete"` here directly?
because if you keep old env key and set its value to "delete" you might face a case that the real-value is delete.  then it will be removed in the ReplaceAll call.
e.g 
```
current := []v1.EnvVar{
		{Name: "env1", Value: "value1"},
		{Name: "env2", Value: "value2"},
		{Name: "env3", Value: "value3"},
		{Name: "env4", Value: "value4"},
	}
	desired := []v1.EnvVar{
		{Name: "env2", Value: "value2"},
		{Name: "env4", Value: "value4"},
	}
```
then you have 
```
expected := []v1.EnvVar{
		{Name: "env2", Value: "value2"},
		{Name: "env4", Value: "value4"},
		{Name: "$patch", Value: "delete"},
		{Name: "$patch", Value: "delete"},
	}
```
and probably can skip `if len(envsToRemove) > 0` check

Reply:
Yeah, I thought about it, but if I do this before marshaling the object `(modJson, err := json.Marshal(deployment))`
The field will be removed as `$patch` is a not valid EnvVar field.

Replacing it after everything was marshaled and the strategic merge patch created was the only way I found to make the strategic merge patch work.

About having the env deleted if its value is "delete," it was intentional
We might need to use a different reserved word for this?

Reply:
sure, my biggest concern is env can be anything set in the deployment by user, and to use a value "delete" is risky. but i guess for now it is only for isvc, it might be ok.

Reply:
> and to use a value "delete" is risky.

yeah, maybe adding another check in the `ReplaceAll` call to make sure there is env to be deleted, mitigate this.

Reply:
in parts

Reply:
or add a placeholder for this case, maybe `mark_for_deletion`

Reply:
Actually, there is a better way to do it, if the env value is "delete" just ignore it.